File: template.go
// This script was written for the advent of code 2024 day X.
// https://adventofcode.com/2025/day/X

package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

const year = 2025

type Solver struct {
	day  int
	year int
}

func NewSolver(day int) *Solver {
	return &Solver{day: day, year: year}
}

func (s *Solver) FetchInput() string {
	inputDir := "input"
	if err := os.MkdirAll(inputDir, 0755); err != nil {
		panic(err)
	}

	filename := filepath.Join(inputDir, fmt.Sprintf("day%02d.txt", s.day))

	if data, err := os.ReadFile(filename); err == nil {
		return strings.TrimRight(string(data), "\n")
	}

	cookie, err := os.ReadFile("cookie.txt")
	if err != nil {
		panic("cookie.txt not found - add your session cookie")
	}

	url := fmt.Sprintf("https://adventofcode.com/%d/day/%d/input", s.year, s.day)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Cookie", fmt.Sprintf("session=%s", strings.TrimSpace(string(cookie))))

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	if err := os.WriteFile(filename, body, 0644); err != nil {
		panic(err)
	}

	return strings.TrimRight(string(body), "\n")
}

func (s *Solver) TestCase(input string, expected1, expected2 any) {
	if expected1 != nil {
		result := s.SolveFirst(input)
		if result == expected1 {
			fmt.Printf("✓ Part 1 test passed: %v\n", result)
		} else {
			fmt.Printf("✗ Part 1 test failed: got %v, expected %v\n", result, expected1)
		}
	}

	if expected2 != nil {
		result := s.SolveSecond(input)
		if result == expected2 {
			fmt.Printf("✓ Part 2 test passed: %v\n", result)
		} else {
			fmt.Printf("✗ Part 2 test failed: got %v, expected %v\n", result, expected2)
		}
	}
}

func (s *Solver) SolveFirst(input string) any {
	lines := strings.Split(input, "\n")
	_ = lines // use lines

	return nil
}

func (s *Solver) SolveSecond(input string) any {
	lines := strings.Split(input, "\n")
	_ = lines // use lines

	return nil
}

// Helper: parse lines into []int
func parseInts(input string) []int {
	var nums []int
	for _, line := range strings.Split(input, "\n") {
		if n, err := strconv.Atoi(line); err == nil {
			nums = append(nums, n)
		}
	}
	return nums
}

// Helper: parse space-separated ints from a line
func parseLineInts(line string) []int {
	var nums []int
	for _, s := range strings.Fields(line) {
		if n, err := strconv.Atoi(s); err == nil {
			nums = append(nums, n)
		}
	}
	return nums
}

func main() {
	solver := NewSolver(1)

	// Test with example input
	testInput := ``
	solver.TestCase(testInput, nil, nil)

	// Solve with real input
	input := solver.FetchInput()
	fmt.Println("Part 1:", solver.SolveFirst(input))
	fmt.Println("Part 2:", solver.SolveSecond(input))
}

File: template.rb
# This script was written for the advent of code 2025 day X.
# https://adventofcode.com/2025/day/X

require 'uri'
require 'net/http'

class Solver
  def initialize(day, year = 2025)
	@day = day
	@year = year
  end

  def fetch_input
	input_directory = 'input'
	unless Dir.exist? input_directory
	  Dir.mkdir input_directory
	end
	
	input_file_name = "#{@year}-#{@day}-input.txt"
	input_file_path = File.join input_directory, input_file_name
	
	if File.exist? input_file_path
	  input_file = File.open input_file_path
	  @input = input_file.read.chomp
	  input_file.close
	  return
	end
	
	auth_filename = 'cookie.txt'
	unless File.exist? auth_filename
	  abort('"cookie.txt" is required to get the puzzle input for your account.')
	end

	cookie_file = File.open auth_filename

	uri = URI("https://adventofcode.com/#{@year}/day/#{@day}/input")
	cookie_value = cookie_file.read.chomp
	user_agent = 'github.com/naturaln0va/aoc2025 by Ryan Ackermann'
	headers = { 'Cookie' => "session=#{cookie_value}", 'User-Agent' => user_agent }
	
	cookie_file.close

	puts "fetching the puzzle input for day #{@day}..."
	res = Net::HTTP.get_response(uri, headers)

	unless res.is_a? Net::HTTPSuccess
	  abort("HTTP Error: #{res.code} - #{res.body}")
	end
	
	@input = res.body
	
	File.write(input_file_path, @input)
  end

  def test_case
	puts "===TEST==="
	first_test_input = ""
	first_answer = solve_first(first_test_input)
	puts "1st answer: #{first_answer}"
	second_test_input = first_test_input
	second_answer = solve_second(second_test_input)
	puts "2nd answer: #{second_answer}"
  end

  def decipher
	fetch_input
	puts "===PUZZLE==="
	first_answer = solve_first(@input)
	puts "1st answer: #{first_answer}"
	second_answer = solve_second(@input)
	puts "2nd answer: #{second_answer}"
  end

  def solve_first(input)
	answer = 0
	lines = input.lines.map(&:strip)
	numbers = input.split().map { |s| s.to_i }
	answer
  end

  def solve_second(input)
	answer = 0
	lines = input.lines.map(&:strip)
	numbers = input.split().map { |s| s.to_i }
	answer
  end
end

s = Solver.new(day = 1)
s.test_case
puts ""
s.decipher